File: sdk/go/worker/subscriber.go
package worker

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"

	amqpadapter "github.com/relaymesh/relaybus/sdk/amqp/go"
	relaymessage "github.com/relaymesh/relaybus/sdk/core/go/message"
	kafkaadapter "github.com/relaymesh/relaybus/sdk/kafka/go"
	natsadapter "github.com/relaymesh/relaybus/sdk/nats/go"
)

// MessageHandler receives relaybus messages for a subscription.
type MessageHandler func(ctx context.Context, msg relaymessage.Message) error

// Subscriber represents a Relaybus subscriber.
type Subscriber interface {
	Start(ctx context.Context, topic string, handler MessageHandler) error
	Close() error
}

// NewFromConfig creates a new worker from a subscriber configuration.
func NewFromConfig(cfg SubscriberConfig, opts ...Option) (*Worker, error) {
	sub, err := BuildSubscriber(cfg)
	if err != nil {
		return nil, err
	}
	opts = append(opts, WithSubscriber(sub))
	return New(opts...), nil
}

// BuildSubscriber creates a Relaybus subscriber from a configuration.
// It can create a single subscriber or a multi-subscriber that combines multiple drivers.
func BuildSubscriber(cfg SubscriberConfig) (Subscriber, error) {
	drivers := uniqueStrings(cfg.Drivers)
	if cfg.Driver != "" {
		drivers = append(drivers, cfg.Driver)
		drivers = uniqueStrings(drivers)
	}
	if len(drivers) == 0 {
		return nil, errors.New("at least one driver is required")
	}
	if len(drivers) == 1 {
		return newSingleSubscriber(cfg, drivers[0])
	}

	subs := make([]namedSubscriber, 0, len(drivers))
	for _, driver := range drivers {
		sub, err := newSingleSubscriber(cfg, driver)
		if err != nil {
			return nil, err
		}
		subs = append(subs, namedSubscriber{driver: driver, sub: sub})
	}

	return &multiSubscriber{
		subscribers: subs,
	}, nil
}

func newSingleSubscriber(cfg SubscriberConfig, driver string) (Subscriber, error) {
	driver = strings.ToLower(strings.TrimSpace(driver))
	if !isSubscriberDriverSupported(driver) {
		return nil, fmt.Errorf("unsupported subscriber driver: %s", driver)
	}
	return &relaybusSubscriber{driver: driver, cfg: cfg}, nil
}

type relaybusSubscriber struct {
	driver string
	cfg    SubscriberConfig
}

func (s *relaybusSubscriber) Start(ctx context.Context, topic string, handler MessageHandler) error {
	if handler == nil {
		return errors.New("handler is required")
	}
	switch s.driver {
	case "amqp":
		if s.cfg.AMQP.URL == "" {
			return errors.New("amqp url is required")
		}
		sub, err := amqpadapter.NewSubscriber(amqpadapter.SubscriberConfig{
			URL:                s.cfg.AMQP.URL,
			Exchange:           s.cfg.AMQP.Exchange,
			RoutingKeyTemplate: s.cfg.AMQP.RoutingKeyTemplate,
			Queue:              s.cfg.AMQP.Exchange,
			AutoAck:            s.cfg.AMQP.AutoAck,
			MaxMessages:        s.cfg.AMQP.MaxMessages,
			Handler:            handler,
		})
		if err != nil {
			return err
		}
		return sub.Start(ctx, topic)
	case "nats":
		if s.cfg.NATS.URL == "" {
			return errors.New("nats url is required")
		}
		sub, err := natsadapter.NewSubscriber(natsadapter.SubscriberConfig{
			URL:           s.cfg.NATS.URL,
			SubjectPrefix: s.cfg.NATS.SubjectPrefix,
			MaxMessages:   s.cfg.NATS.MaxMessages,
			Handler:       handler,
		})
		if err != nil {
			return err
		}
		return sub.Start(ctx, topic)
	case "kafka":
		sub, err := kafkaadapter.NewSubscriber(kafkaadapter.SubscriberConfig{
			Brokers:     s.cfg.Kafka.Brokers,
			Broker:      s.cfg.Kafka.Broker,
			GroupID:     s.cfg.Kafka.GroupID,
			MaxMessages: s.cfg.Kafka.MaxMessages,
			Handler:     handler,
		})
		if err != nil {
			return err
		}
		kafkaTopic := topic
		if s.cfg.Kafka.TopicPrefix != "" {
			kafkaTopic = s.cfg.Kafka.TopicPrefix + topic
		}
		return sub.Start(ctx, kafkaTopic)
	default:
		return fmt.Errorf("unsupported subscriber driver: %s", s.driver)
	}
}

func (s *relaybusSubscriber) Close() error {
	return nil
}

type multiSubscriber struct {
	subscribers []namedSubscriber
}

type namedSubscriber struct {
	driver string
	sub    Subscriber
}

func (m *multiSubscriber) Start(ctx context.Context, topic string, handler MessageHandler) error {
	if len(m.subscribers) == 0 {
		return errors.New("no subscribers configured")
	}
	if handler == nil {
		return errors.New("handler is required")
	}

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	errCh := make(chan error, len(m.subscribers))
	var wg sync.WaitGroup
	wg.Add(len(m.subscribers))

	for _, entry := range m.subscribers {
		entry := entry
		go func() {
			defer wg.Done()
			errCh <- entry.sub.Start(ctx, topic, func(ctx context.Context, msg relaymessage.Message) error {
				if msg.Metadata == nil {
					msg.Metadata = map[string]string{}
				}
				if entry.driver != "" {
					msg.Metadata[MetadataKeyDriver] = entry.driver
				}
				return handler(ctx, msg)
			})
		}()
	}

	go func() {
		wg.Wait()
		close(errCh)
	}()

	var err error
	for subErr := range errCh {
		if subErr != nil {
			err = errors.Join(err, subErr)
			cancel()
		}
	}
	return err
}

func (m *multiSubscriber) Close() error {
	var firstErr error
	for _, entry := range m.subscribers {
		if err := entry.sub.Close(); err != nil && firstErr == nil {
			firstErr = err
		}
	}
	return firstErr
}

func uniqueStrings(values []string) []string {
	seen := make(map[string]struct{}, len(values))
	out := make([]string, 0, len(values))
	for _, value := range values {
		value = strings.ToLower(strings.TrimSpace(value))
		if value == "" {
			continue
		}
		if _, ok := seen[value]; ok {
			continue
		}
		seen[value] = struct{}{}
		out = append(out, value)
	}
	return out
}

func isSubscriberDriverSupported(driver string) bool {
	switch strings.ToLower(driver) {
	case "amqp", "nats", "kafka":
		return true
	default:
		return false
	}
}

File: sdk/typescript/worker/src/subscriber.ts
import { AmqpSubscriber } from "@relaymesh/relaybus-amqp";
import { KafkaSubscriber } from "@relaymesh/relaybus-kafka";
import { NatsSubscriber } from "@relaymesh/relaybus-nats";

import type { Driver, SubscriberConfig } from "./config.js";
import { MetadataKeyDriver } from "./metadata.js";
import type { MessageHandler, RelaybusMessage } from "./types.js";

export interface Subscriber {
  start(topic: string, handler: MessageHandler): Promise<void>;
  close(): Promise<void>;
}

export function buildSubscriber(cfg: SubscriberConfig): Subscriber {
  const drivers = collectDrivers(cfg);
  if (drivers.length === 0) {
    throw new Error("at least one driver is required");
  }
  if (drivers.length === 1) {
    return new RelaybusSubscriber(drivers[0], cfg);
  }

  const entries = drivers.map((driver) => ({
    driver,
    sub: new RelaybusSubscriber(driver, cfg),
  }));
  return new MultiSubscriber(entries);
}

class RelaybusSubscriber implements Subscriber {
  private readonly subs = new Set<{ close: () => Promise<void> }>();

  constructor(private readonly driver: Driver, private readonly cfg: SubscriberConfig) {}

  async start(topic: string, handler: MessageHandler): Promise<void> {
    if (!handler) {
      throw new Error("handler is required");
    }
    switch (this.driver) {
      case "amqp":
        await this.startAmqp(topic, handler);
        return;
      case "kafka":
        await this.startKafka(topic, handler);
        return;
      case "nats":
        await this.startNats(topic, handler);
        return;
      default:
        throw new Error(`unsupported driver: ${this.driver}`);
    }
  }

  async close(): Promise<void> {
    let firstError: Error | undefined;
    for (const sub of this.subs) {
      try {
        await sub.close();
      } catch (err) {
        if (!firstError) {
          firstError = err instanceof Error ? err : new Error(String(err));
        }
      }
    }
    this.subs.clear();
    if (firstError) {
      throw firstError;
    }
  }

  private async startAmqp(topic: string, handler: MessageHandler): Promise<void> {
    const cfg = this.cfg.amqp;
    if (!cfg || !cfg.url) {
      throw new Error("amqp url is required");
    }
    const sub = await AmqpSubscriber.connect({
      url: cfg.url,
      exchange: cfg.exchange,
      routingKeyTemplate: cfg.routingKeyTemplate,
      queue: cfg.queue,
      autoAck: cfg.autoAck,
      maxMessages: cfg.maxMessages,
      onMessage: (msg: RelaybusMessage) => handler(msg),
    });
    this.subs.add(sub);
    try {
      await sub.start(topic);
    } finally {
      this.subs.delete(sub);
    }
  }

  private async startKafka(topic: string, handler: MessageHandler): Promise<void> {
    const cfg = this.cfg.kafka;
    const brokers = cfg?.brokers?.length ? cfg.brokers : cfg?.broker ? [cfg.broker] : [];
    if (!brokers.length) {
      throw new Error("kafka brokers are required");
    }
    const sub = await KafkaSubscriber.connect({
      brokers,
      groupId: cfg?.groupId,
      maxMessages: cfg?.maxMessages,
      onMessage: (msg: RelaybusMessage) => handler(msg),
    });
    this.subs.add(sub);
    try {
      await sub.start(resolveKafkaTopic(topic, cfg));
    } finally {
      this.subs.delete(sub);
    }
  }

  private async startNats(topic: string, handler: MessageHandler): Promise<void> {
    const cfg = this.cfg.nats;
    if (!cfg || !cfg.url) {
      throw new Error("nats url is required");
    }
    const sub = await NatsSubscriber.connect({
      url: cfg.url,
      subjectPrefix: cfg.subjectPrefix,
      maxMessages: cfg.maxMessages,
      onMessage: (msg: RelaybusMessage) => handler(msg),
    });
    this.subs.add(sub);
    try {
      await sub.start(topic);
    } finally {
      this.subs.delete(sub);
    }
  }
}

class MultiSubscriber implements Subscriber {
  constructor(private readonly entries: Array<{ driver: Driver; sub: Subscriber }>) {}

  async start(topic: string, handler: MessageHandler): Promise<void> {
    if (!handler) {
      throw new Error("handler is required");
    }
    await Promise.all(
      this.entries.map((entry) =>
        entry.sub.start(topic, async (msg) => {
          const metadata = { ...(msg.metadata ?? {}) };
          metadata[MetadataKeyDriver] = entry.driver;
          msg.metadata = metadata;
          await handler(msg);
        }),
      ),
    );
  }

  async close(): Promise<void> {
    let firstError: Error | undefined;
    for (const entry of this.entries) {
      try {
        await entry.sub.close();
      } catch (err) {
        if (!firstError) {
          firstError = err instanceof Error ? err : new Error(String(err));
        }
      }
    }
    if (firstError) {
      throw firstError;
    }
  }
}

function collectDrivers(cfg: SubscriberConfig): Driver[] {
  const drivers = new Set<Driver>();
  const append = (value?: Driver) => {
    const normalized = (value ?? "").toLowerCase().trim() as Driver;
    if (normalized) {
      drivers.add(normalized);
    }
  };
  append(cfg.driver);
  if (Array.isArray(cfg.drivers)) {
    for (const driver of cfg.drivers) {
      append(driver);
    }
  }
  return Array.from(drivers);
}

function resolveKafkaTopic(topic: string, cfg?: SubscriberConfig["kafka"]): string {
  const prefix = cfg?.topicPrefix ?? "";
  if (!prefix) {
    return topic;
  }
  return `${prefix}${topic}`;
}
